src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java

```diff
@@ -79,8 +79,10 @@ public SpellResult perform(CastContext context) {
         for (String spellKey : spells) {
             MageSpell spell = mage.getSpell(spellKey);
             long requiredCastCount = spell.getRequiredUpgradeCasts();
+            String requiredPathKey = spell.getRequiredUpgradePath();
             SpellTemplate upgradeSpell = spell.getUpgrade();
-            if (requiredCastCount > 0 && upgradeSpell != null) {
+            Set<String> requiredPathTags = spell.getRequiredUpgradeTags();
+            if (upgradeSpell != null && (requiredCastCount > 0 || requiredPathKey != null || (requiredPathTags != null && !requiredPathTags.isEmpty()))) {
```

Comment:
I *think* all of these checks are unnecessary since spell.getUpgrade performs them too?